### ISSUE TITLE
Improvements for C function names generated by rock

### ISSUE THREAD

**Issue by sebastianbaginski** [magic-lang/ooc-kean#998]
Assume the following code resides in a file called `ClassName.ooc` in the path `name/space`:
```ooc
ClassName: class {
	property: Int { get set }
	init: func { this property = This staticProperty }
	init: func ~overload (value: Int) { this property = value }
	method: func { this property += 1 }
	method: func ~overload (delta: Int) { this property += delta }
	operator + (other: This) { This new(this property + other property) }
	operator + (other: Int) { This new(this property + other) }
	staticProperty: static Int { get set }
	staticMethod: static func { This staticProperty += 1 }
	staticMethod: static func ~overload (delta: Int) { This staticProperty += delta }
}
operator + (left: Int, right: ClassName) { right + left }
function: func -> ClassName { ClassName new() }
function: func ~overload (value: Int) -> ClassName { ClassName new(value) }
```
This can be compiled with:
```bash
rock -v --gc=off --nomain --nolines --driver=make name/space/ClassName.ooc
```

**Comment by simonmika** [comment]
Today this gets named like this in c:
```c
// struct for object
struct _name_space_ClassName__ClassName { ... };
// struct for class of object
struct _name_space_ClassName__ClassNameClass { ... };
// methods & properties
name_space_ClassName__ClassNameClass *name_space_ClassName__ClassName_class();
name_space_ClassName__ClassName* name_space_ClassName__ClassName_new();
name_space_ClassName__ClassName* name_space_ClassName__ClassName_new_overload(lang_Numbers__Int value);
void name_space_ClassName__ClassName_init(name_space_ClassName__ClassName* this);
void name_space_ClassName__ClassName_init_overload(name_space_ClassName__ClassName* this, lang_Numbers__Int value);
void name_space_ClassName__ClassName_method(name_space_ClassName__ClassName* this);
void name_space_ClassName__ClassName_method_impl(name_space_ClassName__ClassName* this);
void name_space_ClassName__ClassName_method_overload(name_space_ClassName__ClassName* this, lang_Numbers__Int delta);
void name_space_ClassName__ClassName_method_overload_impl(name_space_ClassName__ClassName* this, lang_Numbers__Int delta);
void name_space_ClassName__ClassName___OP_ADD_This(name_space_ClassName__ClassName* this, name_space_ClassName__ClassName* other);
void name_space_ClassName__ClassName___OP_ADD_This_impl(name_space_ClassName__ClassName* this, name_space_ClassName__ClassName* other);
void name_space_ClassName__ClassName___OP_ADD_Int(name_space_ClassName__ClassName* this, lang_Numbers__Int other);
void name_space_ClassName__ClassName___OP_ADD_Int_impl(name_space_ClassName__ClassName* this, lang_Numbers__Int other);
void name_space_ClassName__ClassName_staticMethod();
void name_space_ClassName__ClassName_staticMethod_overload(lang_Numbers__Int delta);
lang_Numbers__Int name_space_ClassName__ClassName___getproperty__(name_space_ClassName__ClassName* this);
lang_Numbers__Int name_space_ClassName__ClassName___getproperty___impl(name_space_ClassName__ClassName* this);
void name_space_ClassName__ClassName___setproperty__(name_space_ClassName__ClassName* this, lang_Numbers__Int property);
void name_space_ClassName__ClassName___setproperty___impl(name_space_ClassName__ClassName* this, lang_Numbers__Int property);
void name_space_ClassName__ClassName___defaults__(name_space_ClassName__ClassName* this);
void name_space_ClassName__ClassName___defaults___impl(name_space_ClassName__ClassName* this);
void name_space_ClassName__ClassName___load__();
lang_Numbers__Int name_space_ClassName__ClassName___getstaticProperty__();
void name_space_ClassName__ClassName___setstaticProperty__(lang_Numbers__Int staticProperty);
// static functions
void name_space_ClassName_load();
name_space_ClassName__ClassName* name_space_ClassName__function();
name_space_ClassName__ClassName* name_space_ClassName__function_overload(lang_Numbers__Int value);
void name_space_ClassName____OP_ADD_Int_ClassName(lang_Numbers__Int left, name_space_ClassName__ClassName* right);

```

**Comment by simonmika** [comment]
We want this to instead be:
```c
// struct for object
struct _name_space_ClassName { ... };
// struct for class of object
struct _name_space_ClassNameClass { ... };
// methods & properties
name_space_ClassNameClass *name_space_ClassName_class();
name_space_ClassName* name_space_ClassName_new();
name_space_ClassName* name_space_ClassName_new_overload(int value);
void name_space_ClassName_init(name_space_ClassName* this);
void name_space_ClassName_init_overload(name_space_ClassName* this, int value);
void name_space_ClassName_method(name_space_ClassName* this);
void name_space_ClassName_method_overload(name_space_ClassName* this, int delta);
void name_space_ClassName_ADD_ClassName(name_space_ClassName* this, name_space_ClassName* other);
void name_space_ClassName_ADD_Int(name_space_ClassName* this, int other);
void name_space_ClassName_staticMethod();
void name_space_ClassName_staticMethod_overload(int delta);
int name_space_ClassName_getProperty(name_space_ClassName* this);
void name_space_ClassName_setProperty(name_space_ClassName* this, int property);
int name_space_ClassName_getStaticProperty();
void name_space_ClassName_setStaticProperty(int staticProperty);
// static functions
void name_space_ClassName_load();
name_space_ClassName* name_space_function();
name_space_ClassName* name_space_function_overload(int value);a
void name_space_Int_ADD_ClassName(int left, name_space_ClassName* right);
```

**Comment by sebastianbaginski** [comment]
I think it looks a lot better. @emilwestergren do you have anything to add ? would you prefer to debug new or old version ? ;)

**Comment by sebastianbaginski** [comment]
Of course there will be problem with two files in the same namespace with class `ClassName` in both files if we would like to combine them in one object library. I don't think it is the case in our current source code.

**Comment by emilwestergren** [comment]
Yes it looks much better.

**Comment by simonmika** [comment]
@thomasfanell @davidhesselbom @fredrikbryntesson 
Any comments?

**Comment by davidhesselbom** [comment]
Fine by me.

**Comment by thomasfanell** [comment]
I'm currently investigating this, i'll keep you posted.

**Comment by davidhesselbom** [comment]
You *may* want to have some kind of sanity check, though, just in case someone names two classes in the same namespace (but not in the same `.ooc` file) by the same name. I have no idea what kind of strange compiler errors that would result in.

**Comment by marcusnaslund** [comment]
Having the option (maybe through a flag to `rock`) to be able to generate C code *without* adding line references to `ooc` code would be appreciated by me, at least.

**Comment by thomasfanell** [comment]
> Having the option (maybe through a flag to rock) to be able to generate C code without adding line references to ooc code would be appreciated by me, at least.

What do you mean? This sounds exactly what `--nolines` does.

**Comment by marcusnaslund** [comment]
> What do you mean? This sounds exactly what --nolines does.

I was unaware of that. Thank you. :)

**Comment by thomasfanell** [comment]
[I find browsing this file is much easier than reading the output from rock itself](https://github.com/cogneco/rock/blob/master/source/rock/frontend/Help.ooc)

**Comment by davidpiuva** [comment]
When mangling names, a proof is needed to show that two different identifiers in different name spaces can not have the same mangled name. That is why double underscore and limitations to identifiers is often used. `draw_car_blue` can be either `_car_blue` in the `draw` namespace or `_blue` in the `draw_car` namespace.

**Comment by simonmika** [comment]
I suggest disallowing the use of underscore in identifiers except for the first character. That solvers the issue with needing double underscores. At least almost all problems.

**Comment by davidpiuva** [comment]
Then we have to stick to only separating by single underscore which may collide with functions built in to the language.

**Comment by marcusnaslund** [comment]
Technically not a `kean` issue but a `rock` issue, I've included it with other related issues in #1278 .

**Comment by marcusnaslund** [comment]
This isn't really a ooc-kean/magic-sdk issue, but rather a rock issue. @thomasfanell @sebastianbaginski will you create an issue in the correct repo? I think we all agree that it should look as above.
